File: problems/00030-substring-with-concatenation-of-all-words/go/main.go
package main

import "maps"

func findSubstring(s string, words []string) []int {
	wordCount := map[string]int{}
	syllLen := len(words[0])
	wordLen := len(words) * syllLen
	for _, w := range words {
		wordCount[w]++
	}
	inds := []int{}
	copied := make(map[string]int, len(wordCount))
sIter:
	for i := 0; i <= len(s)-wordLen; i++ {
		maps.Copy(copied, wordCount)
		for j := i; j < i+wordLen; j += syllLen {
			copied[s[j:j+syllLen]]--
			if x := copied[s[j:j+syllLen]]; x < 0 {
				maps.DeleteFunc(copied, func(key string, value int) bool {
					return true
				})
				continue sIter
			}
		}
		inds = append(inds, i)
	}
	return inds
}

File: problems/00030-substring-with-concatenation-of-all-words/rust/src/main.rs
#![allow(dead_code)]
struct Solution {}

use std::collections::HashMap;

impl Solution {
    pub fn find_substring(s: String, words: Vec<String>) -> Vec<i32> {
        let mut result = Vec::with_capacity(s.len());
        let syll_len = words[0].len();
        let word_len = words.len() * syll_len;
        if s.len() < word_len {
            return Vec::new();
        }
        let word_count = words.into_iter().fold(HashMap::new(), |mut map, word| {
            map.entry(word).and_modify(|e| *e += 1).or_insert(1);
            map
        });
        's_iter: for i in 0..=(s.len() - word_len) {
            let mut cloned = word_count.clone();
            for j in (i..(i + word_len)).step_by(syll_len) {
                match cloned.get_mut(&s[j..(j + syll_len)]) {
                    Some(v) if *v > 0 => {
                        *v -= 1;
                    }
                    _ => continue 's_iter,
                }
            }
            result.push(i as i32);
        }
        result
    }
}

fn main() {
    println!("This program doesn't have testing since there's no suitable way of complex testing");
}
